File: 13.romanToInt.go
package main

func romanToInt(s string) int {
	size := len(s)

	sum := 0

	for i := 0; i < size; i++ {
		curr := s[i]
		next := byte(0)
		if i+1 < size {
			next = s[i+1]
		}

		if curr == 'I' {
			if next == 'V' {
				sum += 4
				i++
			} else if next == 'X' {
				sum += 9
				i++
			} else {
				sum += 1
			}
		} else if curr == 'V' {
			sum += 5
		} else if curr == 'X' {
			if next == 'L' {
				sum += 40
				i++
			} else if next == 'C' {
				sum += 90
				i++
			} else {
				sum += 10
			}
		} else if curr == 'L' {
			sum += 50
		} else if curr == 'C' {
			if next == 'D' {
				sum += 400
				i++
			} else if next == 'M' {
				sum += 900
				i++
			} else {
				sum += 100
			}
		} else if curr == 'D' {
			sum += 500
		} else if curr == 'M' {
			sum += 1000
		}
	}

	return sum
}

File: 13.romanToInt.c
#include "shdstd.h"

int romanToInt(char *s)
{
	/*
	 * I -> check next [V,X]
	 * X -> check next [L,C]
	 * C -> check next [D,M]
	 */

	int sum = 0;

	int c;
	int ptr = 0;
	while ((c = s[ptr++]) != '\0') {
		if (c == 'I') {
			if (s[ptr] == 'V') {
				sum += 4;
				++ptr;
			} else if (s[ptr] == 'X') {
				sum += 9;
				++ptr;
			} else {
				sum += 1;
			}
		} else if (c == 'V') {
			sum += 5;
		} else if (c == 'X') {
			if (s[ptr] == 'L') {
				sum += 40;
				++ptr;
			} else if (s[ptr] == 'C') {
				sum += 90;
				++ptr;
			} else {
				sum += 10;
			}
		} else if (c == 'L') {
			sum += 50;
		} else if (c == 'C') {
			if (s[ptr] == 'D') {
				sum += 400;
				++ptr;
			} else if (s[ptr] == 'M') {
				sum += 900;
				++ptr;
			} else {
				sum += 100;
			}
		} else if (c == 'D') {
			sum += 500;
		} else if (c == 'M') {
			sum += 1000;
		}
	}

	return sum;
}
